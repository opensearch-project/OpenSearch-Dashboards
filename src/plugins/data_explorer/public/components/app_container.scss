$osdHeaderOffset: $euiHeaderHeightCompensation;

.deSidebar {
  height: 100%;

  @include ouiBreakpoint("xs", "s", "m") {
    max-width: initial;
    width: 100%;
  }
}

.deLayout {
  height: calc(100vh - #{$osdHeaderOffset});

  &__canvas {
    height: 100%;
  }
}

.headerIsExpanded .deLayout {
  height: calc(100vh - #{$osdHeaderOffset * 2});
}
