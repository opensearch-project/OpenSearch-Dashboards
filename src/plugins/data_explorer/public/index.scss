$osdHeaderOffset: $euiHeaderHeightCompensation;

.dePageTemplate {
  height: calc(100vh - #{$osdHeaderOffset});
}

.headerIsExpanded .dePageTemplate {
  height: calc(100vh - #{$osdHeaderOffset * 2});
}
