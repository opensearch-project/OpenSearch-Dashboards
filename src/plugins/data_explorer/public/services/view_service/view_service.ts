/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Any modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { CoreService } from '../../../../../core/types';
import { ViewDefinition } from './types';
import { View } from './view';

/**
 * Visualization Types Service
 *
 * @internal
 */
export class ViewService implements CoreService<ViewServiceSetup, ViewServiceStart> {
  private views: Record<string, View> = {};

  private registerView(view: View) {
    if (view.id in this.views) {
      throw new Error(`A view with the id ${view.id} already exists!`);
    }
    this.views[view.id] = view;
  }

  public setup() {
    return {
      /**
       * registers a visualization type
       * @param config - visualization type definition
       */
      registerView: <T = any>(config: ViewDefinition<T>): void => {
        const view = new View(config);
        this.registerView(view);
      },
    };
  }

  public start() {
    return {
      /**
       * returns specific View or undefined if not found
       * @param {string} id - id of view to return
       */
      get: (id: string): View | undefined => {
        return this.views[id];
      },
      /**
       * returns all registered Views
       */
      all: (): View[] => {
        return Object.values(this.views);
      },
    };
  }

  public stop() {
    // nothing to do here yet
  }
}

/** @internal */
export type ViewServiceSetup = ReturnType<ViewService['setup']>;
export type ViewServiceStart = ReturnType<ViewService['start']>;
