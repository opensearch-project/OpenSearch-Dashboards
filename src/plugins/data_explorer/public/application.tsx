/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { Provider as ReduxProvider } from 'react-redux';
import { Router, Route, Switch } from 'react-router-dom';
import { AppMountParameters, CoreStart } from '../../../core/public';
import { OpenSearchDashboardsContextProvider } from '../../opensearch_dashboards_react/public';
import { DataExplorerServices } from './types';
import { DataExplorerApp } from './components/app';
import { Store } from './utils/state_management';

export const renderApp = (
  core: CoreStart,
  services: DataExplorerServices,
  params: AppMountParameters,
  store: Store
) => {
  const { history, element } = params;
  ReactDOM.render(
    <Router history={history}>
      <OpenSearchDashboardsContextProvider services={services}>
        <ReduxProvider store={store}>
          <services.i18n.Context>
            <Switch>
              <Route path={[`/:appId`, '/']} exact={false}>
                <DataExplorerApp params={params} />
              </Route>
            </Switch>
          </services.i18n.Context>
        </ReduxProvider>
      </OpenSearchDashboardsContextProvider>
    </Router>,
    element
  );

  return () => ReactDOM.unmountComponentAtNode(element);
};
