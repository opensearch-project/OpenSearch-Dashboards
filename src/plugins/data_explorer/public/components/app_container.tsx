/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { EuiPage, EuiPageBody } from '@elastic/eui';
import { Suspense } from 'react';
import { AppMountParameters } from '../../../../core/public';
import { Sidebar } from './sidebar';
import { NoView } from './no_view';
import { View } from '../services/view_service/view';
import './app_container.scss';

export const AppContainer = ({ view, params }: { view?: View; params: AppMountParameters }) => {
  // TODO: Make this more robust.
  if (!view) {
    return <NoView />;
  }

  const { Canvas, Panel, Context } = view;

  // Render the application DOM.
  return (
    <EuiPage className="deLayout" paddingSize="none">
      {/* TODO: improve fallback state */}
      <Suspense fallback={<div>Loading...</div>}>
        <Context {...params}>
          <Sidebar>
            <Panel {...params} />
          </Sidebar>
          <EuiPageBody className="deLayout__canvas">
            <Canvas {...params} />
          </EuiPageBody>
        </Context>
      </Suspense>
    </EuiPage>
  );
};
