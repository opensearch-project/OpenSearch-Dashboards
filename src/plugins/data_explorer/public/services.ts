/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { createGetterSetter } from '../../opensearch_dashboards_utils/public';
import { UsageCollectionSetup } from '../../usage_collection/public';

export const [getUsageCollector, setUsageCollector] = createGetterSetter<UsageCollectionSetup>(
  'UsageCollector'
);
