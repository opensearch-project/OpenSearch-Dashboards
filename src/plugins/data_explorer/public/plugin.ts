/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { BehaviorSubject } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import {
  AppMountParameters,
  CoreSetup,
  CoreStart,
  Plugin,
  AppNavLinkStatus,
  ScopedHistory,
  AppUpdater,
} from '../../../core/public';
import {
  DataExplorerPluginSetup,
  DataExplorerPluginSetupDependencies,
  DataExplorerPluginStart,
  DataExplorerPluginStartDependencies,
  DataExplorerServices,
} from './types';
import { PLUGIN_ID, PLUGIN_NAME } from '../common';
import { ViewService } from './services/view_service';
import {
  createOsdUrlStateStorage,
  createOsdUrlTracker,
  withNotifyOnErrors,
} from '../../opensearch_dashboards_utils/public';
import { getPreloadedStore } from './utils/state_management';
import { opensearchFilters } from '../../data/public';

export class DataExplorerPlugin
  implements
    Plugin<
      DataExplorerPluginSetup,
      DataExplorerPluginStart,
      DataExplorerPluginSetupDependencies,
      DataExplorerPluginStartDependencies
    > {
  private viewService = new ViewService();
  private appStateUpdater = new BehaviorSubject<AppUpdater>(() => ({}));
  private stopUrlTracking?: () => void;
  private currentHistory?: ScopedHistory;

  public setup(
    core: CoreSetup<DataExplorerPluginStartDependencies, DataExplorerPluginStart>,
    { data }: DataExplorerPluginSetupDependencies
  ): DataExplorerPluginSetup {
    const viewService = this.viewService;

    const { appMounted, appUnMounted, stop: stopUrlTracker } = createOsdUrlTracker({
      baseUrl: core.http.basePath.prepend(`/app/${PLUGIN_ID}`),
      defaultSubUrl: '#/',
      storageKey: `lastUrl:${core.http.basePath.get()}:${PLUGIN_ID}`,
      navLinkUpdater$: this.appStateUpdater,
      toastNotifications: core.notifications.toasts,
      stateParams: [
        {
          osdUrlKey: '_g',
          stateUpdate$: data.query.state$.pipe(
            filter(
              ({ changes }) => !!(changes.globalFilters || changes.time || changes.refreshInterval)
            ),
            map(({ state }) => ({
              ...state,
              filters: state.filters?.filter(opensearchFilters.isFilterPinned),
            }))
          ),
        },
      ],
      getHistory: () => {
        return this.currentHistory!;
      },
    });
    this.stopUrlTracking = () => {
      stopUrlTracker();
    };

    // Register an application into the side navigation menu
    core.application.register({
      id: PLUGIN_ID,
      title: PLUGIN_NAME,
      navLinkStatus: AppNavLinkStatus.hidden,
      mount: async (params: AppMountParameters) => {
        // Load application bundle
        const { renderApp } = await import('./application');

        const [coreStart, pluginsStart] = await core.getStartServices();
        this.currentHistory = params.history;

        // make sure the index pattern list is up to date
        pluginsStart.data.indexPatterns.clearCache();

        const services: DataExplorerServices = {
          ...coreStart,
          scopedHistory: this.currentHistory,
          data: pluginsStart.data,
          embeddable: pluginsStart.embeddable,
          expressions: pluginsStart.expressions,
          osdUrlStateStorage: createOsdUrlStateStorage({
            history: this.currentHistory,
            useHash: coreStart.uiSettings.get('state:storeInSessionStorage'),
            ...withNotifyOnErrors(coreStart.notifications.toasts),
          }),
          viewRegistry: viewService.start(),
        };

        // Get start services as specified in opensearch_dashboards.json
        // Render the application
        const { store, unsubscribe: unsubscribeStore } = await getPreloadedStore(services);
        services.store = store;

        const unmount = renderApp(coreStart, services, params, store);
        appMounted();

        return () => {
          unsubscribeStore();
          appUnMounted();
          unmount();
        };
      },
    });

    return {
      ...this.viewService.setup(),
    };
  }

  public start(core: CoreStart): DataExplorerPluginStart {
    return {};
  }

  public stop() {
    if (this.stopUrlTracking) {
      this.stopUrlTracking();
    }
  }
}
