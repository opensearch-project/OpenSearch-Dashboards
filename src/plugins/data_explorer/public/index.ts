/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import './index.scss';

import { DataExplorerPlugin } from './plugin';

// This exports static code and TypeScript types,
// as well as, OpenSearch Dashboards Platform `plugin()` initializer.
export function plugin() {
  return new DataExplorerPlugin();
}
export { DataExplorerPluginSetup, DataExplorerPluginStart, DataExplorerServices } from './types';
export { ViewProps, ViewDefinition, DefaultViewState } from './services/view_service';
export {
  RootState,
  Store,
  useTypedSelector,
  useTypedDispatch,
  setIndexPattern,
} from './utils/state_management';
