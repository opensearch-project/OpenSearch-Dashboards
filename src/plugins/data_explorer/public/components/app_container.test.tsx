/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { AppContainer } from './app_container';
import { View } from '../services/view_service/view';
import { AppMountParameters } from '../../../../core/public';
import { render } from 'test_utils/testing_lib_helpers';

describe('DataExplorerApp', () => {
  const createView = () => {
    return new View({
      id: 'test-view',
      title: 'Test View',
      defaultPath: '/test-path',
      appExtentions: {} as any,
      Canvas: (() => <div>canvas</div>) as any,
      Panel: (() => <div>panel</div>) as any,
      Context: (() => <div>Context</div>) as any,
    });
  };

  const params: AppMountParameters = {
    element: document.createElement('div'),
    history: {} as any,
    onAppLeave: jest.fn(),
    setHeaderActionMenu: jest.fn(),
    appBasePath: '',
  };

  it('should render NoView when a non existent view is selected', () => {
    const { container } = render(<AppContainer params={params} />);

    expect(container).toContainHTML('View not found');
  });

  it('should render the canvas and panel when selected', () => {
    const view = createView();
    const { container } = render(<AppContainer view={view} params={params} />);

    expect(container).toMatchSnapshot();
  });
});
