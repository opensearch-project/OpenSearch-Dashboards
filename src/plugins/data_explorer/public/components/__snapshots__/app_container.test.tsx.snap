// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataExplorerApp should render the canvas and panel when selected 1`] = `
<div>
  <div
    class="euiPage euiPage--grow deLayout"
  >
    <div>
      Context
    </div>
  </div>
</div>
`;
